ext/compressors/lz4/lz4_compress.c

```diff
@@ -121,40 +136,139 @@ lz4_decompress(WT_COMPRESSOR *compressor, WT_SESSION *session,
+	 */
+	memcpy(&prefix, src, sizeof(LZ4_PREFIX));
+
+	/*
+	 * Decompress, starting after the prefix bytes. Use safe decompression:
+	 * we rely on decompression to detect corruption.
+	 *
+	 * Two code paths, one with and one without a bounce buffer. Our caller
+	 * doesn't store the length of the compressed source and so when using
+	 * raw compression, won't have provided us a large enough buffer, so we
+	 * have to allocate a scratch buffer.
+	 */
+	if (dst_len < prefix.compressed_src) {
+		if ((dst_tmp = wt_api->scr_alloc(
+		   wt_api, session, (size_t)prefix.compressed_src)) == NULL)
+			return (ENOMEM);
+
+		decoded = LZ4_decompress_safe(
+		    (const char *)src + sizeof(LZ4_PREFIX), (char *)dst_tmp,
+		    (int)prefix.compressed_len, (int)prefix.compressed_src);
+
+		if (decoded >= 0)
+			memcpy(dst, dst_tmp, dst_len);
+		wt_api->scr_free(wt_api, session, dst_tmp);
+	} else
+		decoded = LZ4_decompress_safe(
+		    (const char *)src + sizeof(LZ4_PREFIX),
+		    (char *)dst, (int)prefix.compressed_len, (int)dst_len);
+
+	if (decoded >= 0) {
+		*result_lenp = prefix.decompressed_return;
+		return (0);
 	}
 
-	/* Skip over the data size to the start of compressed data. */
-	compressed_data = (char *)src + sizeof(size_t);
+	return (
+	    lz4_error(compressor, session, "LZ4 decompress error", decoded));
+}
+
+/*
+ * lz4_find_slot --
+ *	Find the slot containing the target offset (binary search).
+ */
+static inline uint32_t
+lz4_find_slot(int target_arg, uint32_t *offsets, uint32_t slots)
+{
+	uint32_t base, indx, limit, target;
+
+	indx = 1;					/* -Wuninitialized */
+
+	target = (uint32_t)target_arg;			/* Type conversion */
+
+	/* Fast check if we consumed it all, it's a likely result. */
+	if (target >= offsets[slots])
+		return (slots);
 
 	/*
-	 * The destination buffer length should always be sufficient because
-	 * wiredtiger keeps track of the byte count before compression.  Use
-	 * safe decompression: we may be relying on decompression to detect
-	 * corruption.
+	 * Figure out which slot we got to: binary search. Note the test of
+	 * offset (slot + 1), that's (end-byte + 1) for slot.
 	 */
-	decoded = LZ4_decompress_safe(
-	    compressed_data, (char *)dst, (int)src_data_len, (int)dst_len);
+	for (base = 0, limit = slots; limit != 0; limit >>= 1) {
+		indx = base + (limit >> 1);
+		if (target > offsets[indx + 1]) {
+			base = indx + 1;
+			--limit;
+		}
+	}
+
+	return (indx);
+}
+
+/*
+ * lz4_compress_raw --
+ *	Pack records into a specified on-disk page size.
+ */
+static int
+lz4_compress_raw(WT_COMPRESSOR *compressor, WT_SESSION *session,
+    size_t page_max, int split_pct, size_t extra,
+    uint8_t *src, uint32_t *offsets, uint32_t slots,
+    uint8_t *dst, size_t dst_len, int final,
+    size_t *result_lenp, uint32_t *result_slotsp)
+{
+	LZ4_PREFIX prefix;
+	int lz4_len;
+	uint32_t slot;
+	int sourceSize, targetDestSize;
+
+	(void)compressor;				/* Unused parameters */
+	(void)session;
+	(void)split_pct;
```

Comment:
I've never really understood the intent of `split_pct` -- having a single number suggests that there is some optimal split behavior for a given table that we can know at creation time (even if we do a bulk load followed by general updates).  I suspect that zlib ignoring it contributes to the significantly better compression ratios we get with zlib vs snappy.

I'm inclined to match zlib behavior: as far as I know, we don't have any evidence suggesting that we could do better.

That said, we could change this calculation at any time (and we could consider hacks like having bulk load set `final=1`): it just controls the target page size which won't cause problems with any existing data.